### PR TITLE
CASMINST-5004

### PR DESCRIPTION
**Pull request by leliasen-hpe**
## Summary and Scope

Add storage functionality to Cray-nls. 

File 'storage.rebuild.yaml' is included only for testing purposes

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [issue id](issue link)
* Change will also be needed in `<insert branch name here>`
* Future work required by [issue id](issue link)
* Documentation changes required in [issue id](issue link)
* Merge with/before/after `<insert PR URL here>`

## Testing

_List the environments in which these changes were tested._

### Tested on:

  * `<development system>`
  * Local development environment
  * Virtual Shasta

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)?
- Were continuous integration tests run? If not, why?
- Was upgrade tested? If not, why?
- Was downgrade tested? If not, why?
- Were new tests (or test issues/Jiras) created for this change?

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_

## Pull Request Checklist

- [ ] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [ ] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: api/argo-templates/ncn/storage.rebuild.yaml [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#
```
Comment (dzou-hpe) [inline api/argo-templates/ncn/storage.rebuild.yaml:1]: note that this folder is only used by test. our real workflow should live in docs-csm repo

File: api/argo-templates/render_test.go [hunk 2]
```diff
@@ -101,7 +101,63 @@ func TestRenderWorkerRebuildTemplate(t *testing.T) {
...
+				_, err := GetStorageRebuildWorkflow(rebuildWorkflowFS, req)
+				if (err != nil) != tt.wantErr {
+					t.Errorf("got %v, wantErr %v", err, tt.wantErr)
+					return
+				}
+			})
+		}
+
+	})
+	t.Run("It should select nodes that is not being rebuilt", func(t *testing.T) {
```
Comment (dzou-hpe) [inline api/argo-templates/render_test.go:151]: you won't need this, storage nodes are not part of k8s so argo jobs won't ever run on storage nodes itself

File: utils/env.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ type Env struct {
 	ArgoToken                  string `mapstructure:"ARGO_TOKEN"`
 	ArgoServerURL              string `mapstructure:"ARGO_SERVER_URL"`
 	WorkerRebuildWorkflowFiles string `mapstructure:"WORKER_REBUILD_WORKFLOW_FILES"`
+	StorageRebuildWorkflowFiles string `mapstructure:"WORKER_REBUILD_WORKFLOW_FILES"`
```
Comment (dzou-hpe) [inline utils/env.go:37]: `mapstructure:"WORKER_REBUILD_WORKFLOW_FILES"` should be `mapstructure:"STORAGE_REBUILD_WORKFLOW_FILES"`

File: utils/validator.go [hunk 4]
```diff
@@ -37,24 +37,65 @@ func NewValidator() Validator {
 	return Validator{}
 }
 
+func (validator Validator) ValidateHostnames(hostnames []string) error {
+	var firstHostname = hostnames[0]
+	isWorker, err := regexp.Match(`^ncn-w[0-9]*$`, []byte(firstHostname))
+	if err != nil {
+		return err
+	}
+	if isWorker {
```
Comment (dzou-hpe) [inline utils/validator.go:46]: i don't think this if-else is needed here. what i'd do here:
```
        var firstHostname = hostnames[0]
        // worker
	isWorker, err := regexp.Match(`^ncn-w[0-9]*$`, []byte(firstHostname))
	if err != nil {
		return err
	}
	if isWorker {
	   return validator.ValidateWorkerHostnames(hostnames)
	}
        // storage
        isStorage, err := regexp.Match(`^ncn-s[0-9]*$`, []byte(firstHostname))
        if err != nil {
	        return err
        }
        if isStorage {
	        return validator.ValidateStorageHostnames(hostnames)
        }
        // master/xyz/whateverInTheFuture
        
        // nothing valid so we return error 
        return fmt.Errorf("invalid  hostname: %v", firstHostnames)
```

**Review by dzou-hpe — CHANGES_REQUESTED** [review]

File: api/argo-templates/ncn/storage.rebuild.yaml (hunk 1 above)
Comment (leliasen-hpe) [reply api/argo-templates/ncn/storage.rebuild.yaml:1]: Noted

File: api/argo-templates/render_test.go (hunk 2 above)
Comment (leliasen-hpe) [reply api/argo-templates/render_test.go:151]: removed

File: utils/env.go (hunk 3 above)
Comment (leliasen-hpe) [reply utils/env.go:37]: changed

File: utils/validator.go (hunk 4 above)
Comment (leliasen-hpe) [reply utils/validator.go:46]: good idea, it's changed

Comment (dzou-hpe) [reply utils/validator.go:46]: hmm i don't see the change here, forget to git push?

**Review by dzou-hpe — APPROVED** [review]

**Pull request merged**